plugins/mac-capture/mac-window-capture.m

```diff
@@ -13,71 +13,91 @@
 
 	struct cocoa_window window;
 
-	//CGRect              bounds;
-	//CGWindowListOption  window_option;
 	CGWindowImageOption image_option;
 
-	CGColorSpaceRef color_space;
-
 	DARRAY(uint8_t) buffer;
 
+	uint64_t last_tick;
+	struct obs_source_frame frame;
+
 	pthread_t capture_thread;
 	os_event_t *capture_event;
 	os_event_t *stop_event;
+
+	pthread_mutex_t mutex;
 };
 
 static CGImageRef get_image(struct window_capture *wc)
 {
-	NSArray *arr = (NSArray *)CGWindowListCreate(
-		kCGWindowListOptionIncludingWindow, wc->window.window_id);
-	[arr autorelease];
+	CFMutableArrayRef arr =
+		CFArrayCreateMutable(kCFAllocatorDefault, 0, NULL);
+	CFArrayAppendValue(arr, (void *)(uintptr_t)wc->window.window_id);
+
+	CGImageRef image = CGWindowListCreateImageFromArray(
+		CGRectNull, (CFArrayRef)arr, wc->image_option);
 
-	if (!arr.count && !find_window(&wc->window, NULL, false))
-		return NULL;
+	if (!image && find_window(&wc->window, NULL))
+		image = get_image(wc);
 
-	return CGWindowListCreateImage(CGRectNull,
-				       kCGWindowListOptionIncludingWindow,
-				       wc->window.window_id, wc->image_option);
+	CFRelease(arr);
```

Comment:
I personally find this a bad practice, exactly because it's hard to tell if the author has forgotten.  Putting all of the free/releasing at the end of the function makes it easy to see what's being released, and making sure everything that needed to be released is being released.  If you release it as soon as possible, then you have releases scattered throughout the function which makes spotting a missing release harder.

Reply:
This is obviously a philosophical debate, as long as the release _does_ actually happen. FWIW, the codebase has a preference for "release as soon as possible" (to not hold onto memory unnecessarily longer than required), but I'm not aware of a hard rule about this.